File: ScoreIt/src/main/java/com/sbgapps/scoreit/ScoreFragment.java
/*
 * Copyright (c) 2014 SBG Apps
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.sbgapps.scoreit;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

/**
 * Created by Stéphane on 28/07/2014.
 */
public class ScoreFragment extends Fragment {

    public static final String TAG = ScoreFragment.class.getName();

    private ScoreListFragment mScoreListFragment;
    private ScoreGraphFragment mScoreGraphFragment;
    private HeaderFragment mHeaderFragment;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_score, null);

        if (null == savedInstanceState) {
            mHeaderFragment = new HeaderFragment();
            getChildFragmentManager()
                    .beginTransaction()
                    .replace(R.id.header_container, mHeaderFragment, HeaderFragment.TAG)
                    .commit();

            mScoreListFragment = new ScoreListFragment();
            getChildFragmentManager()
                    .beginTransaction()
                    .replace(R.id.score_container, mScoreListFragment, ScoreListFragment.TAG)
                    .commit();
        } else {
            mHeaderFragment = (HeaderFragment) getChildFragmentManager()
                    .findFragmentByTag(HeaderFragment.TAG);
            mScoreListFragment = (ScoreListFragment) getChildFragmentManager()
                    .findFragmentByTag(ScoreListFragment.TAG);
            mScoreGraphFragment = (ScoreGraphFragment) getChildFragmentManager()
                    .findFragmentByTag(ScoreGraphFragment.TAG);
        }

        return view;
    }

    public ScoreListFragment getScoreListFragment() {
        return mScoreListFragment;
    }

    public ScoreGraphFragment getScoreGraphFragment() {
        return mScoreGraphFragment;
    }

    public HeaderFragment getHeaderFragment() {
        return mHeaderFragment;
    }

    public void loadFragments() {
        mHeaderFragment = new HeaderFragment();
        mScoreListFragment = new ScoreListFragment();

        FragmentTransaction ft = getChildFragmentManager().beginTransaction();
        ft.setCustomAnimations(R.anim.fade_in, R.anim.fade_out);
        ft.replace(R.id.header_container, mHeaderFragment, HeaderFragment.TAG)
                .commit();

        ft = getChildFragmentManager().beginTransaction();
        ft.setCustomAnimations(R.anim.slide_in_down, R.anim.slide_out_up);
        ft.replace(R.id.score_container, mScoreListFragment, ScoreListFragment.TAG)
                .commit();
    }


    public void switchScoreViews() {
        Fragment fragment;
        String tag;

        if (null != mScoreGraphFragment && mScoreGraphFragment.isVisible()) {
            fragment = mScoreListFragment;
            tag = ScoreListFragment.TAG;
        } else if (null == mScoreGraphFragment) {
            fragment = mScoreGraphFragment = new ScoreGraphFragment();
            tag = ScoreGraphFragment.TAG;
        } else {
            fragment = mScoreGraphFragment;
            tag = ScoreGraphFragment.TAG;
        }

        getChildFragmentManager()
                .beginTransaction()
                .setCustomAnimations(
                        R.anim.slide_in_down,
                        R.anim.slide_out_up)
                .replace(R.id.score_container, fragment, tag)
                .commit();
    }

    public void update() {
        if (null != mHeaderFragment)
            mHeaderFragment.update();
        if (null != mScoreListFragment && mScoreListFragment.isVisible())
            mScoreListFragment.update();
        if (null != mScoreGraphFragment && mScoreGraphFragment.isVisible())
            mScoreGraphFragment.update();
    }

    public void closeOpenedItems() {
        if (null != mScoreListFragment)
            mScoreListFragment.closeOpenedItems();
    }
}

File: ui/src/main/java/com/sbgapps/scoreit/ui/view/ScoreFragment.kt
/*
 * Copyright 2018 Stéphane Baiget
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.sbgapps.scoreit.ui.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.sbgapps.scoreit.ui.R
import com.sbgapps.scoreit.ui.base.BaseFragment
import com.sbgapps.scoreit.ui.ext.observe
import com.sbgapps.scoreit.ui.model.Player
import com.sbgapps.scoreit.ui.viewmodel.UniversalViewModel
import com.sbgapps.scoreit.ui.widget.LinearListView
import kotlinx.android.synthetic.main.fragment_score.*
import kotlinx.android.synthetic.main.item_score.view.*
import org.koin.androidx.viewmodel.ext.android.sharedViewModel


class ScoreFragment : BaseFragment() {

    private val model by sharedViewModel<UniversalViewModel>()
    private val adapter = ScoreAdapter()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? = inflater.inflate(R.layout.fragment_score, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        scoreContainer.setAdapter(adapter)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        observe(model.getPlayers(), ::setPlayers)
    }

    private fun setPlayers(players: List<Player>?) {
        players?.let { adapter.items = it }
    }

    inner class ScoreAdapter : LinearListView.Adapter<Player>() {

        override val layoutId = R.layout.item_score

        override fun bind(position: Int, view: View) {
            val player = getItem(position)
            with(view) {
                name.text = player.name
                name.setTextColor(player.color)
                score.text = player.score.toString()
            }
        }
    }

    companion object {
        fun newInstance() = ScoreFragment()
    }
}